File: js/NCList.js
//After
//Alekseyenko, A., and Lee, C. (2007).
//Nested Containment List (NCList): A new algorithm for accelerating
//   interval query of genome alignment and interval databases.
//Bioinformatics, doi:10.1093/bioinformatics/btl647
//http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btl647v1

function NCList() {
}

NCList.prototype.importExisting = function(nclist, attrs, baseURL,
                                           lazyUrlTemplate, lazyClass) {
    this.topList = nclist;
    this.attrs = attrs;
    this.start = attrs.makeFastGetter("Start");
    this.end = attrs.makeFastGetter("End");
    this.lazyClass = lazyClass;
    this.baseURL = baseURL;
    this.lazyUrlTemplate = lazyUrlTemplate;
    this.lazyChunks = {};
};

NCList.prototype.fill = function(intervals, attrs) {
    //intervals: array of arrays of [start, end, ...]
    //attrs: an ArrayRepr object
    //half-open?
    this.attrs = attrs;
    this.start = attrs.makeFastGetter("Start");
    this.end = attrs.makeFastGetter("End");
    var sublist = attrs.makeSetter("Sublist");
    var start = this.start;
    var end = this.end;
    var myIntervals = intervals;
    //sort by OL
    myIntervals.sort(function(a, b) {
        if (start(a) != start(b))
            return start(a) - start(b());
        else
            return end(b) - end(a);
    });
    var sublistStack = new Array();
    var curList = new Array();
    this.topList = curList;
    curList.push(myIntervals[0]);
    var curInterval, topSublist;
    for (var i = 1, len = myIntervals.length; i < len; i++) {
        curInterval = myIntervals[i];
        //if this interval is contained in the previous interval,
        if (end(curInterval) < end(myIntervals[i - 1])) {
            //create a new sublist starting with this interval
            sublistStack.push(curList);
            curList = new Array(curInterval);
            sublist(myIntervals[i - 1], curList);
        } else {
            //find the right sublist for this interval
            while (true) {
                if (0 == sublistStack.length) {
                    curList.push(curInterval);
                    break;
                } else {
                    topSublist = sublistStack[sublistStack.length - 1];
                    if (end(topSublist[topSublist.length - 1])
                        > end(curInterval)) {
                        //curList is the first (deepest) sublist that
                        //curInterval fits into
                        curList.push(curInterval);
                        break;
                    } else {
                        curList = sublistStack.pop();
                    }
                }
            }
        }
    }
};

NCList.prototype.binarySearch = function(arr, item, getter) {
    var low = -1;
    var high = arr.length;
    var mid;

    while (high - low > 1) {
        mid = (low + high) >>> 1;
        if (getter(arr[mid]) > item)
            high = mid;
        else
            low = mid;
    }

    //if we're iterating rightward, return the high index;
    //if leftward, the low index
    if (1 == itemIndex) return high; else return low;
};

NCList.prototype.iterHelper = function(arr, from, to, fun, finish,
                                       inc, searchGet, testGet, path) {
    var len = arr.length;
    var i = this.binarySearch(arr, from, searchGet);
    var getChunk = this.attrs.makeGetter("Chunk");
    var getSublist = this.attrs.makeGetter("Sublist");

    while ((i < len)
           && (i >= 0)
           && ((inc * testGet(arr[i])) < (inc * to)) ) {

        if (arr[i][0] == this.lazyClass) {
            var ncl = this;
            var chunkNum = getChunk(arr[i]);
            if (!(chunkNum in this.lazyChunks)) {
                this.lazyChunks[chunkNum] = {};
            }
            var chunk = this.lazyChunks[chunkNum];
            finish.inc();
            Util.maybeLoad(Util.resolveUrl(this.baseURL,
                                           this.lazyUrlTemplate.replace(
                                                   /\{Chunk\}/g, chunkNum
                                           ) ),
                           chunk,
                           (function (myChunkNum) {
                               return function(o) {
                                   ncl.iterHelper(o, from, to, fun, finish,
                                                  inc, searchGet, testGet,
                                                  [myChunkNum]);
                                   finish.dec();
                               };
                            })(chunkNum),
                           function() {
                               finish.dec();
                           }
                          );

            // // lazy node
            // if (chunk) {
            //     if ("loading" == chunk.state) {
            //         // node is currenly loading; finish this query once it
            //         // has been loaded
            //         finish.inc();
            //         chunk.callbacks.push(
            //             function(parentIndex) {
            //                 return function(o) {
            //                     ncl.iterHelper(o, from, to, fun, finish, inc,
            //                                    searchGet, testGet,
            //                                    [chunkNum]);
            //                     finish.dec();
            //                 };
            //             }(i)
            //         );
            //     } else if ("loaded" == chunk.state) {
            //         this.iterHelper(chunk.sublist, from, to, fun, finish,
            //                         inc, searchGet, testGet, [chunkNum]);
            //     } else {
            //         console.log("unknown lazy type: " + arr[i]);
            //     }
            // } else {
            //     // if the chunk doesn't already exist in this.lazyChunks,
            //     // that means this node hasn't been loaded, so start loading
            //     chunk = {
            //         state: "loading",
            //         callbacks: []
            //     };
            //     this.lazyChunks[chunkNum] = chunk;
            //     finish.inc();
            //     dojo.xhrGet(
            //         {
            //             url: this.baseURL +
            //                 this.lazyUrlTemplate.replace(
            //                         /\{Chunk\}/g, chunkNum
            //                 ),
            //             handleAs: "json",
            //             load: function(myChunk, parentIndex) {
            //                 return function(o) {
            //                     myChunk.state = "loaded";
            //                     myChunk.sublist = o;
            //                     ncl.iterHelper(o, from, to,
            //                                    fun, finish, inc,
            //                                    searchGet, testGet,
            //                                    [chunkNum]);
            //                     for (var c = 0;
            //                          c < lazyObj.callbacks.length;
            //                          c++)
            //                          myChunk.callbacks[c](o);
            //                     finish.dec();
            //                 };
            //             }(chunk, i),
            //             error: function() {
            //                 finish.dec();
            //             }
            //         });
            // }
        } else {
            fun(arr[i], path.concat(i));
        }

        var sublist = getSublist(arr[i]);
        if (sublist)
            this.iterHelper(sublist, from, to,
                            fun, finish, inc, searchGet, testGet,
                            path.concat(i));
        i += inc;
    }
};

NCList.prototype.iterate = function(from, to, fun, postFun) {
    // calls the given function once for each of the
    // intervals that overlap the given interval
    //if from <= to, iterates left-to-right, otherwise iterates right-to-left

    //inc: iterate leftward or rightward
    var inc = (from > to) ? -1 : 1;
    //searchGet: search on start or end
    var searchGet = (from > to) ? this.start : this.end;
    //testGet: test on start or end
    var testIndex = (from > to) ? this.end : this.start;
    var finish = new Finisher(postFun);
    this.iterHelper(this.topList, from, to, fun, finish,
                    inc, searchGet, testGet, [0]);
    finish.finish();
};

NCList.prototype.histogram = function(from, to, numBins, callback) {
    //calls callback with a histogram of the feature density
    //in the given interval

    var result = new Array(numBins);
    var binWidth = (to - from) / numBins;
    var start = this.start;
    var end = this.end;
    for (var i = 0; i < numBins; i++) result[i] = 0;
    //this.histHelper(this.topList, from, to, result, numBins, (to - from) / numBins);
    this.iterate(from, to,
                 function(feat) {
	             var firstBin =
                         Math.max(0, ((start(feat - from) / binWidth) | 0));
                     var lastBin =
                         Math.min(numBins, ((end(feat) - from) / binWidth) | 0);
	             for (var bin = firstBin; bin <= lastBin; bin++)
                         result[bin]++;
                 },
                 function() {
                     callback(result);
                 }
                 );
};

/*

Copyright (c) 2007-2009 The Evolutionary Software Foundation

Created by Mitchell Skinner <mitch_skinner@berkeley.edu>

This package and its accompanying libraries are free software; you can
redistribute it and/or modify it under the terms of the LGPL (either
version 2.1, or at your option, any later version) or the Artistic
License 2.0.  Refer to LICENSE for the full license text.

*/

File: python/nclist.py
#After
#Alekseyenko, A., and Lee, C. (2007).
#Nested Containment List (NCList): A new algorithm for accelerating
#   interval query of genome alignment and interval databases.
#Bioinformatics, doi:10.1093/bioinformatics/btl647
#http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btl647v1

class NCList:
    def __init__(self, start, end, addSublist):
        self.start = start
        self.end = end
        self.addSublist = addSublist
        self.sublistStack = []
        self.count = 0
        self.lastAdded = None
        self.minStart = None
        self.maxEnd = None
        self.curList = []
        self.topList = self.curList
        self.ID = None

    def addFeatures(self, features):
        start = self.start
        end = self.end
        
        if self.lastAdded is None:
            self.lastAdded = features[0]
            features = features[1:]
            self.minStart = start(self.lastAdded)
            self.maxEnd = end(self.lastAdded)
            self.curList.append(self.lastAdded)

        for feat in features:
            # check if the input is sorted by the NCList sort
            # (increasing on start, decreasing on end)
            if ( (start(self.lastAdded) > start(feat))
                 or ( (start(self.lastAdded) == start(feat))
                      and
                      (end(self.lastAdded) < end(feat)) ) ):
                raise InputNotSortedError

            self.maxEnd = max(self.maxEnd, end(feat))
            self.curList = self._addSingle(feat, self.lastAdded,
                                           self.sublistStack, self.curList,
                                           end, self.addSublist)
            self.lastAdded = feat

    def _addSingle(self, feat, lastAdded, sublistStack,
                   curList, end, addSublist):
        # if this interval is contained in the previous interval,
        if end(feat) < end(lastAdded):
            # create a new sublist starting with this interval
            sublistStack.append(curList)
            curList = [feat]
            addSublist(lastAdded, curList)
        else:
            # find the right sublist for this interval
            while True:
                # if we're at the top level list,
                if len(sublistStack) == 0:
                    # just add the current feature to the current list
                    curList.append(feat)
                    break
                else:
                    # if the last interval in the last sublist in sublistStack
                    # ends after the end of the current interval,
                    if end(sublistStack[-1][-1]) > end(feat):
                        # then curList is the first(deepest) sublist
                        # that the current feature fits into, and
                        # we add the current feature to curList
                        curList.append(feat)
                        break
                    else:
                        # move on to the next shallower sublist
                        curList = sublistStack.pop()

        return curList

    @property
    def nestedList(self):
        return self.topList


class LazyNCList:
    def __init__(self, start, end, addSublist, makeLazy,
                 measure, output, sizeThresh):
        self.start = start
        self.end = end
        self.addSublist = addSublist
        self.makeLazy = makeLazy
        self.measure = measure
        self.output = output
        self.sizeThresh = sizeThresh
        self.topList = []
        self.levels = []
        self.chunkNum = 0
        self.lastAdded = None

    def nestedList(self):
        return self.topList

    def addSorted(self, feat):
        start = self.start
        end = self.end

        if self.lastAdded is not None:
            if start(self.lastAdded) > start(feat):
                raise InputNotSortedError
            if ( (start(self.lastAdded) == start(feat))
                 and (end(self.lastAdded) < end(feat)) ):
                raise InputNotSortedError
        self.lastAdded = feat

        featSize = self.measure(feat)
        for level in self.levels:
            level.chunkSize += featSize

            # If:
            #   * this partial chunk is full, or
            #   * this chunk starts at the same place as the feature
            #     immediately before it, and this feature would extend
            #     this chunk beyond that feature, while remaining
            #     within the previous nclist (otherwise we'd try to
            #     add a lazy feature to the previous ncl that doesn't
            #     follow the nclist ordering constraint), or
            #   * this chunk starts at the same place as the previous
            #     nclist and this feature would extend this chunk
            #     beyond that previous nclist (otherwise we'd try to
            #     add a lazy feat to a higher-level ncl that doesn't
            #     follow the nclist ordering constraint)
            if ( (level.chunkSize > self.sizeThresh)
                 or ( (level.prevFeat is not None)
                      and (start(level.prevFeat) == start(level.current[0]))
                      and (end(level.prevFeat) < end(feat))
                      # (level.prevFeat is not None) => (len(level.ncls) > 0)
                      # because the code below sets level.prevFeat and also
                      # calls level.findContainingNcl
                      and (level.ncls[-1].maxEnd >= end(feat)) )
                 or ( (len(level.ncls) > 0)
                      and (level.ncls[-1].minStart == start(level.current[0]))
                      and (level.ncls[-1].maxEnd < end(feat)) ) ):
                # then we're finished with the current "partial" chunk (i.e.,
                # it's now a "complete" chunk rather than a partial one), so
                # create a new NCList to hold all the features in this chunk.
                newNcl = self.makeNcl(level)

                # set the previous feature at this level to the last feature in
                # the partialstack for this level
                level.prevFeat = level.current[-1]

                # start a new partial chunk with the current feature
                level.reset(feat, featSize)
                #level.current = [feat]
                #level.chunkSize = featSize

                # create a lazy ("fake") feature to represent this chunk
                lazyFeat = self.makeLazy(newNcl.minStart, newNcl.maxEnd,
                                         newNcl.ID)

                feat = level.findContainingNcl(self.output, newNcl, lazyFeat)

                # If lazyFeat was contained in a feature in
                # level.ncls, then findContainingNcl will place lazyFeat
                # within that container feature and return None.
                # That means we don't have to proceed to higher levels of the
                # NCL stack to try and find a place to stick feat.
                if feat is None:
                    return

                # if feat is defined, though, then we do have to keep going to
                # find a place for feat
                featSize = self.measure(feat)

            else:
                # we're still filling up the partial chunk at this level, so
                # add the feature there
                level.current.append(feat)
                return

        # if we get through all the levels and still have a feature to place,
        # we create a new highest level and put the feature there
        newToplevel = LazyLevel(feat, featSize)
        #newToplevel.current.append(feat)
        self.levels.append(newToplevel)

    def makeNcl(self, level):
        result = NCList(self.start, self.end, self.addSublist)
        result.ID = self.chunkNum
        self.chunkNum += 1
        result.addFeatures(level.current)
        return result

    def finish(self):
        lazyFeat = None
        for level in self.levels[0:len(self.levels) - 1]:
            if lazyFeat is not None:
                level.current.append(lazyFeat)

            newNcl = self.makeNcl(level)
            lazyFeat = self.makeLazy(newNcl.minStart, newNcl.maxEnd, newNcl.ID)
            # if lazyFeat doesn't get consumed by findContainingNcl, it'll be
            # added to the next highest level on the next loop iteration
            lazyFeat = level.findContainingNcl(self.output, newNcl, lazyFeat)

            for ncl in level.ncls:
                self.output(ncl.nestedList, ncl.ID)

        # make sure there's a top-level NCL
        level = self.levels[-1]
        if lazyFeat is not None:
            level.current.append(lazyFeat)

        newNcl = self.makeNcl(level)
        self.topLevel = newNcl.nestedList


class InputNotSortedError(Exception):
    pass


class LazyLevel:
    def __init__(self, firstFeat, firstSize):
        self.prevFeat = None
        self.current = [firstFeat]
        self.chunkSize = firstSize
        self.ncls = []

    def reset(self, feat, size):
        self.current = [feat]
        self.chunkSize = size

    def findContainingNcl(self, output, newNcl, lazyFeat):
        """
        finds a place in the ncls stack to put the given lazyFeat and newNcl

        takes: function for outputting chunks
               new NCL
               lazy feat for the new NCL
        this sub starts at the end of the array and iterates toward the front,
           examining each NCL in it.  For each of the existing NCLs it
           encounters, it checks to see if that existing NCL contains the new
           NCL.  If it does, then the lazy feat is added to the containing NCL,
           and the new NCL is added to the array; otherwise, the existing NCL
           is popped off of the array, and outputted.
        returns: the lazy feat, if it wasn't consumed by this sub
        """
        while len(self.ncls) > 0:
            existingNcl = self.ncls[-1]
            if newNcl.maxEnd < existingNcl.maxEnd:
                # add the lazy feat to the existing NCL
                existingNcl.addFeatures([lazyFeat])
                # and add the new NCL to the stack
                self.ncls.append(newNcl)
                # and we're done
                return
            else:
                # write out the existing NCL
                output(existingNcl.nestedList, existingNcl.ID)
                self.ncls.pop()

        self.ncls.append(newNcl)
        return lazyFeat
